File: GullyTrends/ViewControllers/SidebarTableViewController.m/SidebarTableViewController.m
//
//  SidebarTableViewController.m
//  GullyTrends
//
//  Created by Lakshmi Vajrapu on 30/10/19.
//  Copyright © 2019 Lakshmi Vajrapu. All rights reserved.
//

#import "SidebarTableViewController.h"
#import "SWRevealViewController.h"
#import "SearchProductViewController.h"
static int const kHeaderSectionTag = 6900;

@interface SidebarTableViewController ()
@property (assign) NSInteger expandedSectionHeaderNumber;
@property (assign) UITableViewHeaderFooterView *expandedSectionHeader;
@property (strong) NSArray *sectionItems;
@property (strong) NSArray *sectionNames;
@property (strong) NSArray *sectionimages;

@property (strong, nonatomic) IBOutlet UITableView *tableView;

@end

@implementation SidebarTableViewController {
    NSArray *menuItems;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.sectionNames = @[ @"Men's", @"Women's", @"Kids",@"Designer Studios",@"Home Shopping", @"Sell On Gully", @"Buy Wholesale", @"Customer Care" ,@"Return Policy", @"FAQS", @"About & Policies"];
    
    self.sectionimages = @[ @"mens-icon", @"womens-icon", @"kids-icon",@"designer-icon",@"menu_home-shopping",@"menu_sell-on-gully-icon",@"menu_buy-wholesale-icon"];

    self.sectionItems = @[ @[@"Shirts", @"T-Shirts", @"Jeans", @"Kurta"],
                           @[@"Jeans", @"Kurta"],
                           @[@"T Shirts", @"Jeans"],
                            @[@"Kurta"]
                         ];
    
    // configure the tableview
    self.tableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    self.tableView.estimatedRowHeight = 100;
    self.expandedSectionHeaderNumber = -1;
    self.tableView.backgroundColor=[UIColor colorWithRed:80.0/255.0 green:24/255.0 blue:133/255.0 alpha:1.0];
    menuItems = @[@"title", @"comments", @"wishlist", @"tag"];
   // [self applyGradientEffect];

}

-(void)applyGradientEffect{
    UIView *view = [[UIView alloc] initWithFrame:CGRectMake(0,0,self.view.bounds.size.width,self.view.bounds.size.height)];
    CAGradientLayer *gradient = [CAGradientLayer layer];
    gradient.frame = view.bounds;
    gradient.startPoint = CGPointZero;
    gradient.endPoint = CGPointMake(0, 1);
    gradient.colors = [NSArray arrayWithObjects:(id)[[UIColor colorWithRed:80/255.0 green:24/255.0 blue:133/255.0 alpha:1.0] CGColor],(id)[[UIColor colorWithRed:14.0/255.0 green:27.0/255.0 blue:96/255.0 alpha:1.0] CGColor], nil];
   // [ self.view.layer addSublayer:gradient];
    
    //# Changed to insertlayer instead addsublayer to fix the issue, after applying gradient labels got disappered.
    [self.view.layer insertSublayer:gradient atIndex:0];

}
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    // Return the number of sections.
    if (self.sectionNames.count > 0) {
        self.tableView.backgroundView = nil;
        return self.sectionNames.count;
    } else {
        UILabel *messageLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, self.view.bounds.size.width, self.view.bounds.size.height)];
        
        messageLabel.text = @"Retrieving data.\nPlease wait.";
        messageLabel.numberOfLines = 0;
        messageLabel.textAlignment = NSTextAlignmentCenter;
        messageLabel.font = [UIFont fontWithName:@"Helvetica Neue" size:20];
        [messageLabel sizeToFit];
        self.tableView.backgroundView = messageLabel;
        
        return 0;
    };
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    // Return the number of rows in the section.
     if (self.expandedSectionHeaderNumber == section) {
           NSMutableArray *arrayOfItems = [self.sectionItems objectAtIndex:section];
           return arrayOfItems.count;
       } else {
           return 0;
       }
}
- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
   // NSString str=[NSString stringWithFormat:@"   %@",[self.sectionNames objectAtIndex:section]];
    if (self.sectionNames.count) {
        return [NSString stringWithFormat:@"         %@",[self.sectionNames objectAtIndex:section]];
    }
    
    return @"";
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section; {
    return 40.0;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    return 35.0;
}

//- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section{
//    UILabel *myLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, tableView.frame.size.width, 30)];
////    myLabel.frame = CGRectMake(20, 8, 320, 20);
//    myLabel.font = [UIFont fontWithName:@"Open Sans-Regular" size:12];
//    myLabel.text = [NSString stringWithFormat:@"   %@",[self.sectionNames objectAtIndex:section]];
//
//    myLabel.backgroundColor=[UIColor blueColor];
//    myLabel.textColor=[UIColor whiteColor];
//    UIView *headerView = [[UIView alloc] init];
//    [headerView addSubview:myLabel];
//    return headerView;
//}

- (void)tableView:(UITableView *)tableView willDisplayHeaderView:(UIView *)view forSection:(NSInteger)section {
    // recast your view as a UITableViewHeaderFooterView
    UITableViewHeaderFooterView *header = (UITableViewHeaderFooterView *)view;
    //header.contentView.backgroundColor = [UIColor blueColor];
   header.contentView.backgroundColor=[UIColor colorWithRed:80.0/255.0 green:24/255.0 blue:133/255.0 alpha:1.0];

    header.textLabel.textColor = [UIColor whiteColor];
    if (section<=6) {
        
    
    UIImageView *viewWithTag = [self.view viewWithTag:kHeaderSectionTag + section];

    if (viewWithTag) {
        [viewWithTag removeFromSuperview];
    }
    // add the arrow image
  //  CGSize headerFrame = self.view.frame.size;
    UIImageView *theImageView = [[UIImageView alloc] initWithFrame:CGRectMake(20, 13, 15, 16)];
    theImageView.image = [UIImage imageNamed:[self.sectionimages objectAtIndex:section]];
    theImageView.tag = kHeaderSectionTag + section;
    theImageView.contentMode = UIViewContentModeScaleAspectFit;
    [header addSubview:theImageView];
    }
    UIImageView *expandimageView = [[UIImageView alloc] initWithFrame:CGRectMake(200, 13, 10, 10)];
    expandimageView.image = [UIImage imageNamed:@"back-arrow"];
//    [header addSubview:expandimageView];

    // make headers touchable
    header.tag = section;
    UITapGestureRecognizer *headerTapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(sectionHeaderWasTouched:)];
    [header addGestureRecognizer:headerTapGesture];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  //  NSString *CellIdentifier = [self.sectionNames objectAtIndex:indexPath.row];
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"title" forIndexPath:indexPath];
    NSArray *section = [self.sectionItems objectAtIndex:indexPath.section];
    cell.contentView.backgroundColor=[UIColor colorWithRed:80.0/255.0 green:24/255.0 blue:133/255.0 alpha:1.0];
    cell.textLabel.textColor=[UIColor whiteColor];
    cell.textLabel.text = [NSString stringWithFormat:@"            %@",[section objectAtIndex:indexPath.row]];

    return cell;
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    
    //self.selectedrow = [self.sectionNames objectAtIndex:indexPath.row];
    [self performSegueWithIdentifier:@"segueID" sender:self];

}


#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    
    // Set the title of navigation bar by using the menu items
    NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
    UINavigationController *destViewController = (UINavigationController*)segue.destinationViewController;
    destViewController.title = [[self.sectionNames objectAtIndex:indexPath.row] capitalizedString];
    
    // Set the photo if it navigates to the PhotoView
    if ([segue.identifier isEqualToString:@"showPhoto"]) {
       // UINavigationController *navController = segue.destinationViewController;
       // SearchProductViewController *productController = [navController childViewControllers].firstObject;
       // NSString *photoFilename = [NSString stringWithFormat:@"%@_photo", [menuItems objectAtIndex:indexPath.row]];
        //photoController.photoFilename = photoFilename;
        UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"product" bundle:nil];
        SearchProductViewController *myNewVC = (SearchProductViewController *)[storyboard instantiateViewControllerWithIdentifier:@"SearchProductViewController"];
        [self.navigationController pushViewController:myNewVC animated:YES];
    }
}
#pragma mark - Expand / Collapse Methods

- (void)sectionHeaderWasTouched:(UITapGestureRecognizer *)sender {
    UITableViewHeaderFooterView *headerView = (UITableViewHeaderFooterView *)sender.view;
    NSInteger section = headerView.tag;
    NSArray *menuitems=@[@"Home Shopping",@"Sell On Gully", @"Buy Wholesale", @"Customer Care" ,@"Return Policy", @"FAQS", @"About & Policies"];
        NSString *searchstring=[headerView.textLabel.text stringByReplacingOccurrencesOfString:@"         " withString:@""];
    if ( [menuitems containsObject:searchstring] ) {
        [self performSegueWithIdentifier:@"segueID" sender:self];
        return;
    }
    UIImageView *eImageView = (UIImageView *)[headerView viewWithTag:kHeaderSectionTag + section];
    self.expandedSectionHeader = headerView;
    
    if (self.expandedSectionHeaderNumber == -1) {
        self.expandedSectionHeaderNumber = section;
        [self tableViewExpandSection:section withImage: eImageView];
    } else {
        if (self.expandedSectionHeaderNumber == section) {
            [self tableViewCollapeSection:section withImage: eImageView];
            self.expandedSectionHeader = nil;
        } else {
            UIImageView *cImageView  = (UIImageView *)[self.view viewWithTag:kHeaderSectionTag + self.expandedSectionHeaderNumber];
            [self tableViewCollapeSection:self.expandedSectionHeaderNumber withImage: cImageView];
            [self tableViewExpandSection:section withImage: eImageView];
        }
    }
}


- (void)tableViewCollapeSection:(NSInteger)section withImage:(UIImageView *)imageView {
    NSArray *sectionData = [self.sectionItems objectAtIndex:section];
    
    self.expandedSectionHeaderNumber = -1;
    if (sectionData.count == 0) {
        return;
    } else {
        [UIView animateWithDuration:0.4 animations:^{
            imageView.transform = CGAffineTransformMakeRotation((0.0 * M_PI) / 180.0);
        }];
        NSMutableArray *arrayOfIndexPaths = [NSMutableArray array];
        for (int i=0; i< sectionData.count; i++) {
            NSIndexPath *index = [NSIndexPath indexPathForRow:i inSection:section];
            [arrayOfIndexPaths addObject:index];
        }
        [self.tableView beginUpdates];
        [self.tableView deleteRowsAtIndexPaths:arrayOfIndexPaths withRowAnimation: UITableViewRowAnimationFade];
        [self.tableView endUpdates];
    }
}

- (void)tableViewExpandSection:(NSInteger)section withImage:(UIImageView *)imageView {
    NSArray *sectionData = [self.sectionItems objectAtIndex:section];
    
    if (sectionData.count == 0) {
        self.expandedSectionHeaderNumber = -1;
        return;
    } else {
        [UIView animateWithDuration:0.4 animations:^{
            imageView.transform = CGAffineTransformMakeRotation((180.0 * M_PI) / 180.0);
        }];
        NSMutableArray *arrayOfIndexPaths = [NSMutableArray array];
        for (int i=0; i< sectionData.count; i++) {
            NSIndexPath *index = [NSIndexPath indexPathForRow:i inSection:section];
            [arrayOfIndexPaths addObject:index];
        }
        self.expandedSectionHeaderNumber = section;
        [self.tableView beginUpdates];
        [self.tableView insertRowsAtIndexPaths:arrayOfIndexPaths withRowAnimation: UITableViewRowAnimationFade];
        [self.tableView endUpdates];
    }
}

@end

File: GullyTrends/ViewControllers/SidebarTableViewController/SidebarTableViewController.swift
//
//  SidebarTableViewController.swift
//  GullyTrends
//
//  Created by karna on 01/01/20.
//  Copyright © 2020 GullyTrends. All rights reserved.
//

import UIKit

private let kHeaderSectionTag = 6900

class SidebarTableViewController: UIViewController {
    private var expandedSectionHeaderNumber = 0
    private var expandedSectionHeader: UITableViewHeaderFooterView?
    private var sectionItems: [[String]]?
    private var sectionNames: [String]?
    //private var onlySectionmenuitems: [String]?
    private var sectionimages: [String]?
    private var pluisMinusLbl: UILabel?
    @IBOutlet private var sideMenuTableView: UITableView!
    private var menuItems: [String]?
    var dataArray = [String: [MenuModel]]()
    lazy var messageLabel: UILabel = {
        let label = UILabel()
        label.frame = CGRect(x: 0, y: 0, width: view.bounds.size.width, height: view.bounds.size.height)
        label.text = "Retrieving data.\nPlease wait."
        label.numberOfLines = 0
        label.textAlignment = .center
        label.font = UIFont(name: "Helvetica Neue", size: 20)
        label.sizeToFit()
        return label
    }()
    override func viewDidLoad() {
        super.viewDidLoad()
        dataArray = SideMenuVM().getMenuData()
        sectionNames = [
        "Men's",
        "Women's",
        "Kids",
        "Designer Studios",
        "Home Shopping",
        "Sell On Gully",
        "Buy Wholesale",
        "Customer Care",
        "Return Policy",
        "FAQS",
        "About & Policies"
        ]

        sectionimages = [
        "mens-icon",
        "womens-icon",
        "kids-icon",
        "designer-icon",
        "menu_home-shopping",
        "menu_sell-on-gully-icon",
        "menu_buy-wholesale-icon"
        ]

        sectionItems = [
            ["Shirts", "T-Shirts", "Jeans", "Kurta"],
            ["Jeans", "Kurta"],
            ["T Shirts", "Jeans"],
            ["Kurta"]
        ]
        
       /* onlySectionmenuitems = [
            "Home Shopping",
            "Sell On Gully",
            "Buy Wholesale",
            "Customer Care",
            "Return Policy",
            "FAQS",
            "About & Policies"
        ]*/
        // configure the tableview
        sideMenuTableView.tableFooterView = UIView(frame: CGRect.zero)
        sideMenuTableView.rowHeight = UITableView.automaticDimension
        sideMenuTableView.estimatedRowHeight = 100
        expandedSectionHeaderNumber = -1
        sideMenuTableView.backgroundColor = UIColor(red: 80.0 / 255.0, green: 24 / 255.0, blue: 133 / 255.0, alpha: 1.0)
        sideMenuTableView.register(SideMenuViewCell.self, forCellReuseIdentifier: SideMenuViewCell.reuseIdentifier)
        menuItems = ["title", "comments", "wishlist", "tag"]
        // [self applyGradientEffect];
        // Do any additional setup after loading the view.
    }
    //    func applyGradientEffect() {
    //        let view = UIView(frame: CGRect(x: 0, y: 0, width: self.view.bounds.size.width, height: self.view.bounds.size.height))
    //        let gradient = CAGradientLayer()
    //        gradient.frame = view.bounds
    //        gradient.startPoint = CGPoint.zero
    //        gradient.endPoint = CGPoint(x: 0, y: 1)
    //        gradient.colors = [(UIColor(red: 80 / 255.0, green: 24 / 255.0, blue: 133 / 255.0, alpha: 1.0)).cgColor, (UIColor(red: 14.0 / 255.0, green: 27.0 / 255.0, blue: 96 / 255.0, alpha: 1.0)).cgColor].compactMap { $0 }
    //        self.view.layer.insertSublayer(gradient, at: 0)
    //    }
    
    
    /*
     // MARK: - Navigation
     
     // In a storyboard-based application, you will often want to do a little preparation before navigation
     override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
     // Get the new view controller using segue.destination.
     // Pass the selected object to the new view controller.
     }
     */
    
}
extension SidebarTableViewController: UITableViewDelegate, UITableViewDataSource {
    func numberOfSections(in tableView: UITableView) -> Int {
        // Return the number of sections.
        if sectionNames?.count ?? 0 > 0 {
            self.sideMenuTableView.backgroundView = nil
            return sectionNames?.count ?? 0
        } else {
          
            self.sideMenuTableView.backgroundView = messageLabel

            return 0
        }
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        // Return the number of rows in the section.
        if expandedSectionHeaderNumber == section {
            let arrayOfItems = sectionItems?[section]
            return arrayOfItems?.count ?? 0
        } else {
            return 0
        }
    }

    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        // NSString str=[NSString stringWithFormat:@"   %@",[self.sectionNames objectAtIndex:section]];
        if ((sectionNames?.count) != nil) {
            return "\(sectionNames?[section] ?? "")"
        }
        return ""
    }
    func tableView(_ tableView: UITableView, willDisplayHeaderView view: UIView, forSection section: Int) {
        let header = view as? UITableViewHeaderFooterView
        header?.contentView.backgroundColor = UIColor(red: 80.0 / 255.0, green: 24 / 255.0, blue: 133 / 255.0, alpha: 1.0)

        header?.textLabel?.textColor = UIColor.white
        if section <= 6 {
            let viewWithTag = self.view.viewWithTag(kHeaderSectionTag + section) as? UIImageView

            if viewWithTag != nil {
                viewWithTag?.removeFromSuperview()
            }
            let theImageView = UIImageView(frame: CGRect(x: 20, y: 13, width: 15, height: 16))
            theImageView.image = UIImage(named: sectionimages?[section] ?? "")
            theImageView.tag = kHeaderSectionTag + section
            theImageView.contentMode = .scaleAspectFit
            header?.addSubview(theImageView)
        }

        //let searchstring = header?.textLabel?.text?.replacingOccurrences(of: "         ", with: "")
        //if !(onlySectionmenuitems?.contains(searchstring ?? "") ?? false) {
         //   pluisMinusLbl = UILabel(frame: CGRect(x: 220, y: 8, width: 25, height: 25))
         //   pluisMinusLbl?.text = "+"
         //   pluisMinusLbl?.textColor = UIColor.white
         //   header?.addSubview(pluisMinusLbl ?? UILabel())
       // }
        header?.tag = section
        let headerTapGesture = UITapGestureRecognizer(target: self, action: #selector(sectionHeaderWasTouched(_:)))
        header?.addGestureRecognizer(headerTapGesture)
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "title", for: indexPath) as! SideMenuViewCell
        let section = sectionItems?[indexPath.section]
        cell.configure(text: section?[indexPath.row] ?? "")
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
//        selectedrow = sectionNames[indexPath.row]
//        performSegue(withIdentifier: "segueID", sender: self)
    }
    
    @objc func sectionHeaderWasTouched(_ sender: UITapGestureRecognizer?) {
        let headerView = sender?.view as? UITableViewHeaderFooterView
        let section = headerView?.tag ?? 0

//        let searchstring = headerView?.textLabel?.text?.replacingOccurrences(of: "         ", with: "")
//        if onlySectionmenuitems?.contains(searchstring ?? "") ?? false {
////            performSegue(withIdentifier: "segueID", sender: self)
//            return
//        }
        let eImageView = headerView?.viewWithTag(kHeaderSectionTag + section) as? UIImageView
        expandedSectionHeader = headerView


        if expandedSectionHeaderNumber == -1 {
            expandedSectionHeaderNumber = section
            tableViewExpandSection(section, withImage: eImageView)
        } else {
            if expandedSectionHeaderNumber == section {
                tableViewCollapeSection(section, withImage: eImageView)
                expandedSectionHeader = nil
            } else {
                let cImageView = view.viewWithTag(kHeaderSectionTag + expandedSectionHeaderNumber) as? UIImageView
                tableViewCollapeSection(expandedSectionHeaderNumber, withImage: cImageView)
                tableViewExpandSection(section, withImage: eImageView)
            }
        }
    }

    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {

        // Set the title of navigation bar by using the menu items
        let indexPath = sideMenuTableView.indexPathForSelectedRow
        let destViewController = segue.destination as? UINavigationController
        destViewController?.title = (sectionNames?[indexPath?.row ?? 0])?.capitalized

        // Set the photo if it navigates to the PhotoView
//        if (segue.identifier == "segueID") {
//            let searchVC = SearchProductViewController()
//            navigationController?.pushViewController(searchVC, animated: true)
//        }
    }
    func tableViewCollapeSection(_ section: Int, withImage imageView: UIImageView?) {
        let sectionData = sectionItems?[section]

        expandedSectionHeaderNumber = -1
        if sectionData?.count == 0 {
            return
        } else {
            UIView.animate(withDuration: 0.4, animations: {
                imageView?.transform = CGAffineTransform(rotationAngle: (0.0 * .pi) / 180.0)
            })
            var arrayOfIndexPaths: [AnyHashable] = []
            for i in 0..<(sectionData?.count ?? 0) {
                let index = IndexPath(row: i, section: section)
                arrayOfIndexPaths.append(index)
            }
            sideMenuTableView.beginUpdates()
            if let arrayOfIndexPaths = arrayOfIndexPaths as? [IndexPath] {
                sideMenuTableView.deleteRows(at: arrayOfIndexPaths, with: .fade)
            }
            sideMenuTableView.endUpdates()
        }
    }

    func tableViewExpandSection(_ section: Int, withImage imageView: UIImageView?) {
        let sectionData = sectionItems?[section]

        if sectionData?.count == 0 {
            expandedSectionHeaderNumber = -1
            return
        } else {
            UIView.animate(withDuration: 0.4, animations: {
                imageView?.transform = CGAffineTransform(rotationAngle: (180.0 * .pi) / 180.0)
            })
            var arrayOfIndexPaths: [AnyHashable] = []
            for i in 0..<(sectionData?.count ?? 0) {
                let index = IndexPath(row: i, section: section)
                arrayOfIndexPaths.append(index)
            }
            expandedSectionHeaderNumber = section
            sideMenuTableView.beginUpdates()
            if let arrayOfIndexPaths = arrayOfIndexPaths as? [IndexPath] {
                sideMenuTableView.insertRows(at: arrayOfIndexPaths, with: .fade)
            }
            sideMenuTableView.endUpdates()
        }
    }

}
